feat: implement 0x35 - CALLDATALOAD Opcode

Quentash opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Resolves: #121

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

Eikix commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,6 +50,42 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+
+            // Append if not out of bounds
+            if i + offset < calldata_len {
+                data_to_load += (*calldata[i + offset]).into();
```
at some point it'd be nice to have u256::from(slice of bytes)

Quentash replied:
You mean like a new helper function ?
Like u256::from(offset) => returns the u256 bytes from the offset-th most value byte ?
Or like u256::from(offset,len) => returns len u256 bytes from the offset-th most value byte ?

enitrat replied:
I think the behavior implemented here is just `load_word` from `utils::helpers`. If so we can just replace that part.

Eikix reviewed (commented):
hi! welcome and thanks for this first pr 🙏🙏
will review a second time tomorrow

Quentash commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,6 +50,42 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
     /// # Specification: https://www.evm.codes/#35?fork=shanghai
     fn exec_calldataload(ref self: ExecutionContext) -> Result<(), EVMError> {
+        // Stack input:
+        // 0 - offset: calldata offset of the word we read (32 byte steps).
```
That was copy/pasted from the 0.X version, but I will remove the () as it brings confusion.
"calldata offset of the word we read" is explanatory engough but I guessed the () wanted to say we read 32 bytes every calldataload call

Quentash commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,6 +50,42 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
     fn exec_calldataload(ref self: ExecutionContext) -> Result<(), EVMError> {
+        // Stack input:
+        // 0 - offset: calldata offset of the word we read (32 byte steps).
+        let offset = self.stack.pop()?.try_into().unwrap();
```
I could implement a U256ToU32 function in helpers and having it handling a new custom error "Kakarot: into() overflow" or whatever you'd prefer ?

enitrat commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,7 +51,41 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
-        Result::Ok(())
+        // Stack input:
+        // 0 - offset: calldata offset of the word we read
+        let offset: u32 = u256_to_u32(self.stack.pop()?)?;
```
for consistency this should be done in the same way as in https://github.com/kkrt-labs/kakarot-ssj/pull/197

enitrat commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,7 +51,41 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+        // Early return if the offset is beyond calldata
+        if offset >= calldata_len.into() {
+            self.stack.push(0);
+            return Result::Ok(());
```
```rust
            return self.stack.push(0)
```

enitrat commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,7 +51,41 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+
+        let mut i: usize = 1;
+        loop {
+            if i > 31 {
```
use equality checks instead of range checks, it's cheaper in Cairo (because the instruction set is limited and doesn't contain native range checks )

```suggestion
            if i == 32 {
```

enitrat commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,7 +51,41 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+            }
+
+            // Shift left
+            data_to_load *= 256;
```
note - related to https://github.com/kkrt-labs/kakarot-ssj/issues/185 for future reference

enitrat replied:
@quentash any risk of this operation overflowing? In that case, do we want to wrap around? If yes you can import `u256_overflow_mul` from `integer` and use it instead

Quentash replied:
No, since we do the operation is always and only made 31 times, it would never overflow.

enitrat commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,7 +51,41 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+            data_to_load *= 256;
+
+            // Append if not out of bounds
+            if i + offset < calldata_len {
```
imo `i` should be initialized at offset directly to make it simpler

enitrat commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -41,3 +43,117 @@ fn test__exec_callvalue() {
+
+    let mut i: u32 = 0;
+    loop {
+        if i > 31 {
```
```suggestion
        if i == 32 {
```

not very important given that it's a test but it's a good habit. We know that `i` is always initialized under the limit and increased only by 1

enitrat commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -41,3 +43,117 @@ fn test__exec_callvalue() {
+            break;
+        }
+
+        if i + offset < call_data_len {
```
actually, instead of doing this, you could just slice call_data starting from`offset` for 32 elements, and use `SpanPartialEq` to compare both arrays.

enitrat replied:
> verify that len isn't >= 32

I would rather not if that condition is never triggered because we don't call that function with len >=32 in the codebase - given that this function interacts with memory we want it as optimized as possible


> modify load_word for it to accepts an offset

you wouldn't need to, you could just do something like

```rust
 let bytes_len = cmp::min(32, calldata_len - offset);
 let slice = calldata.slice(offset, bytes_len);
```
and then call the load_word function which would result in the same thing as starting with an offset.

the only problem remaining here is appending it with zeroes. Which we can be easily overcomed.

Here is my proposal

```rust

        // Copy len_bytes from calldata
        let len_bytes = cmp::min(32, calldata_len - offset);
        let slice = calldata.slice(offset, len_bytes);
        let mut loaded_data = load_word(len_bytes, slice);

        // fill the rest of the word with zeros
       //TODO: optimize once we have dw-based exponentiation
        let mut i = 32 - len_bytes;
        loop {
            if i == 0 {
                break;
            }
            loaded_data *= 256;
            i -= 1;
        };

        self.stack.push(loaded_data)
  ```


fwiw this is equivalent to doing `loaded_data *= 256.pow(i)` - but it's slightly more expensive

enitrat commented on crates/evm/src/instructions/environmental_information.cairo:
```diff
@@ -50,7 +51,41 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
+        let calldata = self.call_context().call_data();
+        let calldata_len = calldata.len();
+
+        // Early return if the offset is beyond calldata
```
```suggestion
        // All bytes after the end of the calldata are set to 0.
```

ClementWalter commented:
> fwiw this is equivalent to doing `loaded_data *= 256.pow(i)` - but it's slightly more expensive 

are you saying that native `core::integers::u256::pow` is more expensive than this loop? is it because it makes some extra checks?
In terms of effectivness, should we instead 
- pad first the `sliced` array with `[0, ..]`
- do `load_word`, and no `loop` nor `pow`

ClementWalter commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+fn test_calldata_load() {
+    // Given
+    let mut ctx = setup_execution_context();
+    let call_data: Span<u8> = ctx.call_context().call_data();
```
it's `calldata` and not `call_data`

ClementWalter commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+    // Given
+    let mut ctx = setup_execution_context();
+    let call_data: Span<u8> = ctx.call_context().call_data();
+    let call_data_len = call_data.len();
```
here as well

ClementWalter commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+
+    let mut expected = load_word(call_data_len, call_data);
+
+    let mut i = 32 - call_data_len;
```
you use the same logic in the function and in the test!

I guess that here it'd be better to use hard coded expexted results, there are fyi 2 examples in the playgroud

ClementWalter commented on crates/utils/Scarb.toml:
```diff
@@ -5,3 +5,4 @@ version = "0.1.0"
 # See more keys and their definitions at https://docs.swmansion.com/scarb/docs/reference/manifest.html
 
 [dependencies]
+evm = { path = "../evm" }
```
why did you add this?

Quentash replied:
Probably forgot to remove it after tests. Now removed!

enitrat commented:
> are you saying that native core::integers::u256::pow is more expensive than this loop? is it because it makes some extra checks?

I'm not sure where you got this one, it doesn't exist and this is our own implementation of `pow` in `utils::math`. But yes, it's more expensive.

> In terms of effectivness, should we instead

pad first the sliced array with [0, ..]
do load_word, and no loop nor pow

will check

enitrat commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_calldata_load() {
```
```suggestion
fn test_calldataload() {
```

enitrat commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_calldata_load_with_offset() {
```
```suggestion
fn test_calldataload_with_offset() {
```

enitrat commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_calldata_load_with_offset_beyond_calldata() {
```
```suggestion
fn test_calldataload_with_offset_beyond_calldata() {
```

enitrat commented on crates/evm/src/tests/test_instructions/test_environment_information.cairo:
```diff
@@ -196,3 +196,111 @@ fn test_returndatasize() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_calldata_load_with_offset_conversion_error() {
```
```suggestion
fn test_calldataload_with_offset_conversion_error() {
```

Quentash commented:
Corrected all the mentioned mistakes.
Added hardcoded expected results.
Added create_context_with_calldata() helper function to be able to feed desired data into the context calldata.
Removed scarb.toml unnecessary entry.

ClementWalter commented:
minor changes plz

Quentash commented:
Minor changes done.

enitrat reviewed (approved):
lgtm
